File: packages/ruby/test/storage_contract_test.rb
require_relative "../lib/identifold/storage"

class FakeStorage
  def reserve(_request) = true

  def resolve(_reference, namespace)
    Identifold::Storage::ReferenceMapping.new(
      machine_id: "01890f8c-7b2a-7cc3-98b0-112233445566",
      namespace: namespace
    )
  end

  def allocate(_request) = 1
end

adapter = Identifold::Storage::Adapter.new(
  reference_store: FakeStorage.new,
  reference_lookup: FakeStorage.new,
  sequence_allocator: FakeStorage.new
)
reservation = Identifold::Storage::ReferenceReservation.new(
  machine_id: "01890f8c-7b2a-7cc3-98b0-112233445566",
  namespace: "order",
  reference: "ORD-0123-4567-89-P"
)
raise "reserve" unless adapter.reference_store.reserve(reservation)
raise "resolve" unless adapter.reference_lookup.resolve(
  reservation.reference, reservation.namespace
).machine_id == reservation.machine_id
raise "allocate" unless adapter.sequence_allocator.allocate(
  Identifold::Storage::SequenceAllocationRequest.new(
    machine_id: reservation.machine_id,
    namespace: "receipt",
    reference_prefix: "RCT",
    scope: nil,
    width: 4
  )
) == 1

File: packages/php/tests/StorageContractTest.php
<?php

declare(strict_types=1);

spl_autoload_register(static function (string $class): void {
    $prefix = 'Greyfoundry\\Identifold\\';
    if (!str_starts_with($class, $prefix)) {
        return;
    }
    $path = __DIR__ . '/../src/' . str_replace('\\', '/', substr($class, strlen($prefix))) . '.php';
    if (is_file($path)) {
        require_once $path;
    }
});

use Greyfoundry\Identifold\Storage\IdentifoldStorageAdapter;
use Greyfoundry\Identifold\Storage\ReferenceLookup;
use Greyfoundry\Identifold\Storage\ReferenceMapping;
use Greyfoundry\Identifold\Storage\ReferenceReservation;
use Greyfoundry\Identifold\Storage\ReferenceStore;
use Greyfoundry\Identifold\Storage\SequenceAllocationRequest;
use Greyfoundry\Identifold\Storage\SequenceAllocator;

$fake = new class () implements ReferenceStore, ReferenceLookup, SequenceAllocator {
    public function reserve(ReferenceReservation $request): bool
    {
        return true;
    }

    public function resolve(string $reference, string $namespace): ?ReferenceMapping
    {
        return new ReferenceMapping('01890f8c-7b2a-7cc3-98b0-112233445566', $namespace);
    }

    public function allocate(SequenceAllocationRequest $request): int
    {
        return 1;
    }
};
$adapter = new IdentifoldStorageAdapter($fake, $fake, $fake);
$reservation = new ReferenceReservation(
    '01890f8c-7b2a-7cc3-98b0-112233445566',
    'order',
    'ORD-0123-4567-89-P',
);
assert($adapter->referenceStore->reserve($reservation));
assert($adapter->referenceLookup->resolve($reservation->reference, $reservation->namespace)?->machineId
    === $reservation->machineId);
assert($adapter->sequenceAllocator->allocate(new SequenceAllocationRequest(
    $reservation->machineId,
    'receipt',
    'RCT',
    null,
    4,
)) === 1);
